File: acwing/算法提高课/第一章动态规划/1024.装箱问题/1024.装箱问题.cpp
#include <algorithm>
#include <iostream>
#include <cstring>

using namespace std;

const int N = 35, V = 20000 + 10;

int f[N][V];
int a[N];

int main()
{
    int v, n;
    cin >> v >> n;
    for (int i = 1; i <= n; i++)
    {
        cin >> a[i];
        for (int j = 1; j <= v; j++)
        {
            f[i][j] = f[i - 1][j];
            if (j >= a[i])
            {
                f[i][j] = max(f[i][j], f[i - 1][j - a[i]] + a[i]);
            }
        }
    }
    cout << (v - f[n][v]) << endl;
    return 0;
}

File: acwing/算法提高课/第一章动态规划/1024.装箱问题/1024.装箱问题.py
V = 20000 + 10
N = 35

f = [[0 for _ in range(V)]for _ in range(N)]

a = []

v = int(input())
n = int(input())

a.append(0)
for _ in range(n):
    a.append(int(input()))

for i in range(1, n + 1):
    for j in range(1, v + 1):
        f[i][j] = f[i - 1][j]
        if j >= a[i]:
            f[i][j] = max(f[i][j], f[i - 1][j - a[i]] + a[i])

print(v - f[n][v])
